File: go/test/project_column_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestProjectColumnEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ProjectColumn(nil)
		if ent == nil {
			t.Fatal("expected non-nil ProjectColumnEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"project_column": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.ProjectColumn(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.ProjectColumn(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := project_columnBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "project_column." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_PROJECT_COLUMN_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		projectColumnRef01Ent := client.ProjectColumn(nil)
		projectColumnRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "project_column"}), "project_column_ref01"))
		projectColumnRef01Data["project_id"] = setup.idmap["project01"]

		projectColumnRef01DataResult, err := projectColumnRef01Ent.Create(projectColumnRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		projectColumnRef01Data = core.ToMapAny(entityData(projectColumnRef01DataResult))
		if projectColumnRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if projectColumnRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		projectColumnRef01Match := map[string]any{
			"project_id": setup.idmap["project01"],
		}

		projectColumnRef01ListResult, err := projectColumnRef01Ent.List(projectColumnRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		projectColumnRef01List, projectColumnRef01ListOk := projectColumnRef01ListResult.([]any)
		if !projectColumnRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", projectColumnRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(projectColumnRef01List), map[string]any{"id": projectColumnRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		projectColumnRef01DataUp0Up := map[string]any{
			"id": projectColumnRef01Data["id"],
		}

		projectColumnRef01MarkdefUp0Name := "cards_url"
		projectColumnRef01MarkdefUp0Value := fmt.Sprintf("Mark01-project_column_ref01_%d", setup.now)
		projectColumnRef01DataUp0Up[projectColumnRef01MarkdefUp0Name] = projectColumnRef01MarkdefUp0Value

		projectColumnRef01ResdataUp0Result, err := projectColumnRef01Ent.Update(projectColumnRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		projectColumnRef01ResdataUp0 := core.ToMapAny(entityData(projectColumnRef01ResdataUp0Result))
		if projectColumnRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if projectColumnRef01ResdataUp0["id"] != projectColumnRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if projectColumnRef01ResdataUp0[projectColumnRef01MarkdefUp0Name] != projectColumnRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", projectColumnRef01MarkdefUp0Name, projectColumnRef01ResdataUp0[projectColumnRef01MarkdefUp0Name])
		}

		// LOAD
		projectColumnRef01MatchDt0 := map[string]any{
			"id": projectColumnRef01Data["id"],
		}
		projectColumnRef01DataDt0Loaded, err := projectColumnRef01Ent.Load(projectColumnRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		projectColumnRef01DataDt0LoadResult := core.ToMapAny(entityData(projectColumnRef01DataDt0Loaded))
		if projectColumnRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if projectColumnRef01DataDt0LoadResult["id"] != projectColumnRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func project_columnBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "project_column", "ProjectColumnTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read project_column test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse project_column test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"project_column01", "project_column02", "project_column03", "project01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_PROJECT_COLUMN_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_PROJECT_COLUMN_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
		"GITHUB_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_PROJECT_COLUMN_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["GITHUB_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/project_column_entity_test.lua
-- ProjectColumn entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("github_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("ProjectColumnEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:ProjectColumn(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["project_column"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:ProjectColumn(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:ProjectColumn(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = project_column_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "project_column." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_PROJECT_COLUMN_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local project_column_ref01_ent = client:ProjectColumn(nil)
    local project_column_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.project_column"), "project_column_ref01"))
    project_column_ref01_data["project_id"] = setup.idmap["project01"]

    local project_column_ref01_data_result, err = project_column_ref01_ent:create(project_column_ref01_data, nil)
    assert.is_nil(err)
    project_column_ref01_data = helpers.to_map(type(project_column_ref01_data_result) == 'table' and project_column_ref01_data_result.data_get and project_column_ref01_data_result:data_get() or project_column_ref01_data_result)
    assert.is_not_nil(project_column_ref01_data)
    assert.is_not_nil(project_column_ref01_data["id"])

    -- LIST
    local project_column_ref01_match = {
      ["project_id"] = setup.idmap["project01"],
    }

    local project_column_ref01_list_result, err = project_column_ref01_ent:list(project_column_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(project_column_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(project_column_ref01_list_result),
      { id = project_column_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local project_column_ref01_data_up0_up = {
      id = project_column_ref01_data["id"],
    }

    local project_column_ref01_markdef_up0_name = "cards_url"
    local project_column_ref01_markdef_up0_value = "Mark01-project_column_ref01_" .. tostring(setup.now)
    project_column_ref01_data_up0_up[project_column_ref01_markdef_up0_name] = project_column_ref01_markdef_up0_value

    local project_column_ref01_resdata_up0_result, err = project_column_ref01_ent:update(project_column_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local project_column_ref01_resdata_up0 = helpers.to_map(type(project_column_ref01_resdata_up0_result) == 'table' and project_column_ref01_resdata_up0_result.data_get and project_column_ref01_resdata_up0_result:data_get() or project_column_ref01_resdata_up0_result)
    assert.is_not_nil(project_column_ref01_resdata_up0)
    assert.are.equal(project_column_ref01_resdata_up0["id"], project_column_ref01_data_up0_up["id"])
    assert.are.equal(project_column_ref01_resdata_up0[project_column_ref01_markdef_up0_name], project_column_ref01_markdef_up0_value)

    -- LOAD
    local project_column_ref01_match_dt0 = {
      id = project_column_ref01_data["id"],
    }
    local project_column_ref01_data_dt0_loaded, err = project_column_ref01_ent:load(project_column_ref01_match_dt0, nil)
    assert.is_nil(err)
    local project_column_ref01_data_dt0_load_result = helpers.to_map(type(project_column_ref01_data_dt0_loaded) == 'table' and project_column_ref01_data_dt0_loaded.data_get and project_column_ref01_data_dt0_loaded:data_get() or project_column_ref01_data_dt0_loaded)
    assert.is_not_nil(project_column_ref01_data_dt0_load_result)
    assert.are.equal(project_column_ref01_data_dt0_load_result["id"], project_column_ref01_data["id"])

  end)
end)

function project_column_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/project_column/ProjectColumnTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read project_column test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "project_column01", "project_column02", "project_column03", "project01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("GITHUB_TEST_PROJECT_COLUMN_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["GITHUB_TEST_PROJECT_COLUMN_ENTID"] = idmap,
    ["GITHUB_TEST_LIVE"] = "FALSE",
    ["GITHUB_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["GITHUB_TEST_PROJECT_COLUMN_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["GITHUB_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["GITHUB_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["GITHUB_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
